src/common/librlist/rnode.c

```diff
@@ -12,72 +12,436 @@
+                                  struct idset *ids)
+{
+    struct rnode *n = rnode_new (name, rank);
+    if (n == NULL)
+        return NULL;
+    if (!(n->cores = rnode_add_child_idset (n, "core", ids, ids)))
         goto fail;
-    n->up = true;
     return (n);
 fail:
     rnode_destroy (n);
     return NULL;
 }
 
+struct rnode *rnode_create_count (const char *name,
+                                  uint32_t rank,
+                                  int count)
+{
+    struct rnode *n = NULL;
+    struct idset *ids = NULL;
+
+    if (!(ids = idset_create (0, IDSET_FLAG_AUTOGROW))
+        || (idset_range_set (ids, 0, count-1) < 0))
+        goto out;
+    n = rnode_create_idset (name, rank, ids);
+out:
+    idset_destroy (ids);
+    return n;
+}
+
+static int rnode_set_empty (struct rnode *n)
+{
+    int count = 0;
+    struct rnode_child *c = zhashx_first (n->children);
+    while (c) {
+        idset_destroy (c->avail);
+        if (!(c->avail = idset_copy (c->ids)))
+            return -1;
+        count += idset_count (c->ids);
+        c = zhashx_next (n->children);
+    }
+    return count;
+}
+
+static int rnode_set_alloc (struct rnode *n)
+{
+    int count = 0;
+    struct rnode_child *c;
+
+    if (!n)
+        return -1;
+
+    c = zhashx_first (n->children);
+    while (c) {
+        if (idset_subtract (c->ids, c->avail) < 0)
+            return -1;
+        idset_destroy (c->avail);
+        if (!(c->avail = idset_copy (c->ids)))
+            return -1;
+        count += idset_count (c->ids);
+        c = zhashx_next (n->children);
+    }
+    return count;
+}
+
+static int rnode_set_avail (struct rnode *n)
+{
+    int count = 0;
+    struct rnode_child *c;
+
+    if (!n)
+        return -1;
+
+    c = zhashx_first (n->children);
+    while (c) {
+        idset_destroy (c->ids);
+        if (!(c->ids = idset_copy (c->avail)))
+            return -1;
+        count += idset_count (c->ids);
+        c = zhashx_next (n->children);
+    }
+    return count;
+}
+
+static zhashx_t *rnode_children_copy (const struct rnode *n)
+{
+    zhashx_t *copy;
+
+    /*  Temporarily set duplicator on orig.
+     *  (if duplicator permanently set then zhashx_insert() also
+     *   duplicates items, which we don't want
+     */
+    zhashx_set_duplicator (n->children,
+                           (zhashx_duplicator_fn *) rnode_child_copy);
+    copy = zhashx_dup (n->children);
+    zhashx_set_destructor (copy, rn_child_free);
+    zhashx_set_duplicator (copy, NULL);
+    zhashx_set_duplicator (n->children, NULL);
+    return copy;
+}
+
+struct rnode *rnode_copy (const struct rnode *orig, int flags)
+{
+    struct rnode *n = rnode_new (orig->hostname, orig->rank);
+    if (!n)
+        return NULL;
+    zhashx_destroy (&n->children);
+    if (!(n->children = rnode_children_copy (orig)))
+        goto fail;
+    if (!(n->cores = zhashx_lookup (n->children, "core")))
+        goto fail;
+    if (flags == RNODE_COPY_ALLOC && rnode_set_alloc (n) <= 0)
```

Comment:
to me `flags` suggests bitwise AND here not `==`.  Should `flags` be an enum type?

Reply:
yeah, I wanted to leave it open to flags in the future. What the function is doing under the covers should be hidden from the api user.